### PR TITLE
Changed inputs.py and date_picker.ts to allow custom date formats but running into error (WIP)

### LINKED ISSUE #10619: [FEATURE] Allow custom date formats in DatePicker (flatpickr)
As-per the Discourse discussion here: https://discourse.bokeh.org/t/flatpickr-date-display-format/6630

Currently,  we are constrained to the ISO yyyy-mm-dd format for the flatpickr implementation of DatePicker (post-2.0);

image

It would be great to expose the existing flatpickr dateFormat options to the user, retaining the standard ISO display as a default.

Looks like flatpickr supports everything from the JS side already via their config.dateFormat options;

https://flatpickr.js.org/formatting/

### PR DESCRIPTION
**Pull request by wenjaylfw**
Changed files that are mentioned in the title but getting error "TypeError: Cannot read property 'set_value' of undefined" when trying to pass in an argument for date format.

A couple of extra files for debugging included that I will remove before final pull request.

- [ ] issues: fixes #10619

### CODE REVIEW

File: bokehjs/src/lib/models/widgets/date_picker.ts [hunk 1]
```diff
@@ -116,6 +119,7 @@ export class DatePicker extends InputWidget {
         enabled_dates:  [ DatesList, [] ],
         position:       [ CalendarPosition, "auto" ],
         inline:         [ Boolean, false ],
+        date_format:    [ String ],
```
Comment (mattpap) [inline bokehjs/src/lib/models/widgets/date_picker.ts:122]: What's the default format?

File: .vscode/launch.json [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    "configurations": [
+        {
+            "type": "chrome",
+            "request": "launch",
+            "name": "Launch Chrome against localhost",
+            "url": "file:///C:/Users/wenja/Documents/VS%20Code/bokeh/tests/unit/bokeh/models/widgets/test_DatePicker.html",
+            "webRoot": "${workspaceFolder}"
+        }
+    ]
+}
```
Comment (mattpap) [inline .vscode/launch.json:15]: This file needs to be removed.

File: tests/unit/bokeh/models/widgets/test_DatePicker.html [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+              });
+            };
+            if (document.readyState != "loading") fn();
+            else document.addEventListener("DOMContentLoaded", fn);
+          })();
+        </script>
+    
+  </body>
+  
+</html>
```
Comment (mattpap) [inline tests/unit/bokeh/models/widgets/test_DatePicker.html:86]: This file needs to be removed.

File: bokehjs/src/lib/models/widgets/date_picker.ts (hunk 1 above)
Comment (philippjfr) [inline bokehjs/src/lib/models/widgets/date_picker.ts:122]:
```suggestion
        date_format:    [ String, "Y-m-d" ],
```

File: examples/app/sliders.py [hunk 4]
```diff
@@ -29,7 +29,7 @@
 
 
 # Set up plot
-plot = figure(plot_height=400, plot_width=400, title="my sine wave",
+plot = figure(plot_height=400, plot_width=400, title="Example for Homework 6a",
```
Comment (philippjfr) [inline examples/app/sliders.py:32]:
```suggestion
plot = figure(plot_height=400, plot_width=400, title="my sine wave",
```

File: bokeh/models/widgets/inputs.py [hunk 5]
```diff
@@ -418,6 +418,13 @@ class DatePicker(InputWidget):
     Whether the calendar sholud be displayed inline.
     """)
 
+    date_format = String(default="Y-m-d", help="""
+    The desired date format to display on the Calendar when a date is picked.
+
+    .. note::
+        Must contain characters from flatpickr's table of Formatting Tokens
```
Comment (bryevdv) [inline bokeh/models/widgets/inputs.py:425]: We should drop an explicit link here to make it easier for users find the table (or just copy it)

File: tests/unit/bokeh/models/widgets/test_DatePicker.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+# Setup
+#-----------------------------------------------------------------------------
+
+#-----------------------------------------------------------------------------
+# General API
+#-----------------------------------------------------------------------------
+
+
+class TestDatePicker:
+    def test_setting_value_as_date_time(self) -> None:
```
Comment (bryevdv) [inline tests/unit/bokeh/models/widgets/test_DatePicker.py:30]: I'm not sure "as_date_time" is an accurate description here, suggest "as_iso_date_format" instead

File: tests/unit/bokeh/models/widgets/test_slider.py [hunk 7]
```diff
@@ -1,4 +1,4 @@
-#-----------------------------------------------------------------------------
+#   -----------------------------------------------------------------------------
```
Comment (bryevdv) [inline tests/unit/bokeh/models/widgets/test_slider.py:1]: this formatting change should be reverted

**Comment by bryevdv** [comment]
@wenjaylfw just a few final comments

**Comment by bryevdv** [comment]
@wenjaylfw back from holidays, I was about to merge this but then I encountered a slight confusion about this test:
```
    def test_value_in_custom_format(self) -> None:
        set_date = "2019-02-01"
        s = DatePicker(value="02 01 2019", date_format="m d Y")
        assert s.value == (set_date)
```
I thought the intention of `date_format` was only to affect how the dates are *displayed* by the picker, so I guess I'd expect all data values to still be set via an ISO date format (and so a test for this would really need to examine the DOM to verify the correct format was displayed). I am actually a little surprised that setting a value like `"02 01 2019"` passes validation on the `Date` property.

**Comment by bryevdv** [comment]
@wenjaylfw I am going to go ahead and close this for lack of activity. If you are still interested in working on the display of dates in teh date picker we can reopen or open a new PR

**Pull request closed without merging**